File: app/src/main/java/com/beautifulquran/data/QfMushafMapper.kt
package com.beautifulquran.data

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put

/** Joins QF's three independent resources onto our canonical Quran word rows. */
internal fun mapQfMushaf(
    canonical: Map<Int, Map<Int, List<String>>>,
    mushafRows: List<QfCacheRow>,
    translationRows: List<QfCacheRow>,
    transliterationRows: List<QfCacheRow>,
    supplementRows: List<QfCacheRow>,
    expectedPages: IntRange = 1..604,
    json: Json = Json { ignoreUnknownKeys = true },
): List<QfCacheRow> {
    val mushaf = mushafRows.map { json.parseToJsonElement(it.payload).jsonObject }
    val metadata = mushaf.single { it.string("record_type") == "mushaf" }
    check(metadata.int("pages_count") == expectedPages.count() && metadata.int("lines_per_page") == 15) {
        "Unexpected QF Mushaf layout"
    }
    val pages = mushaf.filter { it.string("record_type") == "mushaf_page" }
    check(pages.size == expectedPages.count() && pages.map { it.int("page_number") }.toSet() == expectedPages.toSet()) {
        "QF Mushaf omitted a page"
    }

    val translations = uniqueWordText(translationRows, json, "translation")
    val transliterations = groupedWordText(transliterationRows, json)
    val supplements = uniqueWordText(supplementRows, json, "supplement")
    val qcfByVerse = mushaf
        .filter { it.string("record_type") == "mushaf_word" }
        .groupBy { it.int("verse_id") }

    val verses = canonical.toSortedMap().flatMap { (surah, ayahs) ->
        ayahs.toSortedMap().map { (ayah, words) -> CanonicalVerse(surah, ayah, words) }
    }
    check(qcfByVerse.keys == (1..verses.size).toSet()) {
        "QF Mushaf verse coverage mismatch"
    }

    val output = buildList {
        verses.forEachIndexed { index, verse ->
            val records = qcfByVerse.getValue(index + 1)
            val words = records.filter { it.string("char_type_name") == "word" }
                .sortedBy { it.int("position_in_verse") }
            val ends = records.filter { it.string("char_type_name") == "end" }
            check(words.indices.all { words[it].int("position_in_verse") == it + 1 } && ends.size == 1) {
                "Invalid QF word topology ${verse.key}"
            }
            addAll(
                mapVerse(
                    verse,
                    words,
                    ends.single().requiredString("text"),
                    translations,
                    transliterations,
                    supplements,
                ),
            )
        }
    }
    check(output.size == verses.sumOf { it.words.size } && output.map { it.recordKey }.distinct().size == output.size) {
        "QF reader view is incomplete"
    }
    assertQcfV2Runs(output, expectedPages)
    return output
}

private fun mapVerse(
    verse: CanonicalVerse,
    qfWords: List<JsonObject>,
    endGlyph: String,
    translations: Map<Long, String>,
    transliterations: Map<Long, List<String>>,
    supplements: Map<Long, String>,
): List<QfCacheRow> {
    val rule = TOPOLOGY_RULES[verse.key]
    val expectedQfCount = verse.words.size + (rule?.qfCount ?: 1) - (rule?.canonicalCount ?: 1)
    check(qfWords.size == expectedQfCount) { "QF topology changed for ${verse.key}" }
    val mapped = mutableMapOf<Int, AlignedQfWord>()
    var canonicalPosition = 1
    var qfPosition = 1
    while (canonicalPosition <= verse.words.size) {
        val atRule = rule?.takeIf {
            it.canonicalStart == canonicalPosition && it.qfStart == qfPosition
        }
        val canonicalCount = atRule?.canonicalCount ?: 1
        val qfCount = atRule?.qfCount ?: 1
        val source = qfWords.subList(qfPosition - 1, qfPosition - 1 + qfCount)
        val page = source.first().int("page_number")
        val line = source.first().int("line_number")
        check(source.all { it.int("page_number") == page && it.int("line_number") == line }) {
            "A joined QF word crosses a line at ${verse.key}"
        }
        val ids = source.map { it.long("word_id") }
        mapped[canonicalPosition] = AlignedQfWord(
            glyph = source.joinToString(" ") { it.requiredString("text") },
            page = page,
            line = line,
            spanEnd = canonicalPosition + canonicalCount - 1,
            translation = joinQfGloss(ids.map(translations::getValue)),
            transliteration = joinQfGloss(ids.map { id ->
                supplements[id] ?: transliterations[id]?.singleOrNull()
                ?: error("QF transliteration is ambiguous or missing for word $id")
            }),
        )
        canonicalPosition += canonicalCount
        qfPosition += qfCount
    }
    check(qfPosition == qfWords.size + 1) { "QF topology ended early for ${verse.key}" }
    val lastOwner = mapped.keys.maxOrNull() ?: error("QF verse ${verse.key} has no words")
    mapped[lastOwner] = mapped.getValue(lastOwner).let { it.copy(glyph = "${it.glyph} $endGlyph") }
    val ayahPage = qfWords.first().int("page_number")

    return verse.words.indices.map { index ->
        val position = index + 1
        val value = mapped[position] ?: AlignedQfWord("", 0, 0, position, "", "")
        val key = "${verse.key}:$position"
        val payload = buildJsonObject {
            put("record_type", "mushaf_word")
            put("record_key", key)
            put("surah_id", verse.surah)
            put("ayah_number", verse.ayah)
            put("position", position)
            put("translation_en", value.translation)
            put("transliteration", value.transliteration)
            put("qcf_v2", value.glyph)
            put("qcf_page", value.page)
            put("qcf_line", value.line)
            put("qcf_span_end", value.spanEnd)
            put("ayah_page", ayahPage)
        }
        QfCacheRow(QF_MUSHAF_RESOURCE, "mushaf_word", key, payload.toString(), "")
    }
}

private fun groupedWordText(rows: List<QfCacheRow>, json: Json): Map<Long, List<String>> =
    rows.map { json.parseToJsonElement(it.payload).jsonObject }
        .groupBy({ it.long("word_id") }, { it.requiredString("text").trim() })

private fun uniqueWordText(
    rows: List<QfCacheRow>,
    json: Json,
    label: String,
): Map<Long, String> {
    val values = groupedWordText(rows, json)
    check(values.values.all { it.size == 1 }) { "Duplicate QF $label word owner" }
    return values.mapValues { it.value.single() }
}

private fun joinQfGloss(parts: List<String>) =
    parts.filter(String::isNotBlank).filterIndexed { index, part -> index == 0 || part != parts[index - 1] }
        .joinToString(" ")

private data class CanonicalVerse(val surah: Int, val ayah: Int, val words: List<String>) {
    val key = "$surah:$ayah"
}

private data class AlignedQfWord(
    val glyph: String,
    val page: Int,
    val line: Int,
    val spanEnd: Int,
    val translation: String,
    val transliteration: String,
)

private data class TopologyRule(
    val canonicalStart: Int,
    val canonicalCount: Int,
    val qfStart: Int,
    val qfCount: Int,
)

/** Ten QAC/QCF token-boundary differences; counts changing makes the cache fail closed. */
private val TOPOLOGY_RULES = mapOf(
    "2:72" to TopologyRule(4, 2, 4, 1),
    "2:181" to TopologyRule(3, 2, 3, 1),
    "8:6" to TopologyRule(4, 2, 4, 1),
    "13:37" to TopologyRule(8, 2, 8, 1),
    "15:7" to TopologyRule(1, 1, 1, 2),
    "27:20" to TopologyRule(4, 1, 4, 2),
    "36:22" to TopologyRule(1, 1, 1, 2),
    "37:130" to TopologyRule(3, 2, 3, 1),
    "37:164" to TopologyRule(1, 1, 1, 2),
    "41:47" to TopologyRule(25, 1, 25, 2),
)

private fun JsonObject.string(name: String) = get(name)?.jsonPrimitive?.contentOrNull
private fun JsonObject.requiredString(name: String) = string(name) ?: error("QF field $name is missing")
private fun JsonObject.int(name: String) = get(name)?.jsonPrimitive?.intOrNull ?: error("QF field $name is missing")
private fun JsonObject.long(name: String) = get(name)?.jsonPrimitive?.longOrNull ?: error("QF field $name is missing")

/** Proves that every glyph is in the contiguous run encoded by its page font. */
internal fun assertQcfV2Runs(rows: List<QfCacheRow>, expectedPages: IntRange) {
    val pages = rows.asSequence()
        .map { Json.parseToJsonElement(it.payload).jsonObject }
        .filter { it.requiredString("qcf_v2").isNotBlank() }
        .groupBy { it.int("qcf_page") }
    expectedPages.forEach { page ->
        val glyphs = pages[page] ?: error("QCF V2 page $page has no glyphs")
        val codes = glyphs
            .sortedWith(compareBy({ it.int("surah_id") }, { it.int("ayah_number") }, { it.int("position") }))
            .flatMap { row ->
                row.requiredString("qcf_v2").codePoints().toArray()
                    .filterNot(Character::isWhitespace).asIterable()
            }
        codes.forEachIndexed { index, code ->
            val expected = QCF_V2_FIRST_CODEPOINT + index
            check(code == expected) {
                "QCF V2 page $page glyph $index is U+${code.toString(16).uppercase()}, " +
                    "expected U+${expected.toString(16).uppercase()}"
            }
        }
    }
}

internal fun readCanonicalWords(database: QuranDatabase): Map<Int, Map<Int, List<String>>> {
    val chapters = mutableMapOf<Int, MutableMap<Int, MutableList<String>>>()
    database.db.rawQuery(
        "SELECT surah_id,ayah_number,arabic FROM words ORDER BY surah_id,ayah_number,position",
        null,
    ).use { cursor ->
        while (cursor.moveToNext()) {
            chapters.getOrPut(cursor.getInt(0)) { mutableMapOf() }
                .getOrPut(cursor.getInt(1)) { mutableListOf() }
                .add(cursor.getString(2))
        }
    }
    return chapters.mapValues { (_, ayahs) -> ayahs.mapValues { it.value.toList() } }
}

private const val QCF_V2_FIRST_CODEPOINT = 0xFC41

File: web/src/data/qfMushafMapper.ts
export interface RuntimeMushafWord {
  record_type: 'mushaf_word'
  record_key: string
  surah_id: number
  ayah_number: number
  position: number
  translation_en: string
  transliteration: string
  qcf_v2: string
  qcf_page: number
  qcf_line: number
  qcf_span_end: number
  ayah_page: number
}

export interface StoredQfResource {
  resourceGroup: string
  resourceId: number
  records: Record<string, unknown>[]
}

type TopologyRule = {
  canonicalStart: number
  canonicalCount: number
  qfStart: number
  qfCount: number
}

const TOPOLOGY_RULES = new Map<string, TopologyRule>([
  ['2:72', { canonicalStart: 4, canonicalCount: 2, qfStart: 4, qfCount: 1 }],
  ['2:181', { canonicalStart: 3, canonicalCount: 2, qfStart: 3, qfCount: 1 }],
  ['8:6', { canonicalStart: 4, canonicalCount: 2, qfStart: 4, qfCount: 1 }],
  ['13:37', { canonicalStart: 8, canonicalCount: 2, qfStart: 8, qfCount: 1 }],
  ['15:7', { canonicalStart: 1, canonicalCount: 1, qfStart: 1, qfCount: 2 }],
  ['27:20', { canonicalStart: 4, canonicalCount: 1, qfStart: 4, qfCount: 2 }],
  ['36:22', { canonicalStart: 1, canonicalCount: 1, qfStart: 1, qfCount: 2 }],
  ['37:130', { canonicalStart: 3, canonicalCount: 2, qfStart: 3, qfCount: 1 }],
  ['37:164', { canonicalStart: 1, canonicalCount: 1, qfStart: 1, qfCount: 2 }],
  ['41:47', { canonicalStart: 25, canonicalCount: 1, qfStart: 25, qfCount: 2 }],
])

/** Join QF's independently synced resources into one fully validated reader view. */
export function normalizeQfMushaf(
  canonical: Map<string, string[]>,
  resources: StoredQfResource[],
  expectedPages: readonly number[],
): RuntimeMushafWord[] {
  const mushaf = resource(resources, 'mushafs', 1).records
  const metadata = mushaf.filter((row) => row.record_type === 'mushaf')
  if (metadata.length !== 1 || Number(metadata[0]!.pages_count) !== expectedPages.length ||
      Number(metadata[0]!.lines_per_page) !== 15) throw new Error('Unexpected QF Mushaf layout')
  const pages = mushaf.filter((row) => row.record_type === 'mushaf_page')
  if (pages.length !== expectedPages.length ||
      new Set(pages.map((row) => Number(row.page_number))).size !== expectedPages.length ||
      expectedPages.some((page) => !pages.some((row) => Number(row.page_number) === page))) {
    throw new Error('QF Mushaf omitted a page')
  }

  const translations = uniqueWordText(resource(resources, 'word_by_word_translations', 59).records, 'translation')
  const transliterations = groupedWordText(resource(resources, 'word_by_word_transliterations', 60).records)
  const supplements = uniqueWordText(resource(resources, 'word_supplements', 1).records, 'supplement')
  const byVerse = new Map<number, Record<string, unknown>[]>()
  for (const row of mushaf.filter((value) => value.record_type === 'mushaf_word')) {
    const verseId = integer(row.verse_id, 'verse_id')
    const rows = byVerse.get(verseId) ?? []
    rows.push(row)
    byVerse.set(verseId, rows)
  }

  const verses = [...canonical.entries()].sort(([a], [b]) => compareVerseKeys(a, b))
  if (byVerse.size !== verses.length || verses.some((_, index) => !byVerse.has(index + 1))) {
    throw new Error('QF Mushaf verse coverage mismatch')
  }
  const output: RuntimeMushafWord[] = []
  verses.forEach(([verseKey, canonicalWords], index) => {
    const qfRows = byVerse.get(index + 1)!
    const words = qfRows.filter((row) => row.char_type_name === 'word')
      .sort((a, b) => Number(a.position_in_verse) - Number(b.position_in_verse))
    const ends = qfRows.filter((row) => row.char_type_name === 'end')
    if (words.some((word, position) => Number(word.position_in_verse) !== position + 1) || ends.length !== 1) {
      throw new Error(`Invalid QF word topology ${verseKey}`)
    }
    output.push(...mapVerse(
      verseKey, canonicalWords, words, String(ends[0]!.text),
      translations, transliterations, supplements,
    ))
  })
  const canonicalCount = [...canonical.values()].reduce((sum, words) => sum + words.length, 0)
  if (output.length !== canonicalCount || new Set(output.map((row) => row.record_key)).size !== output.length) {
    throw new Error('QF reader view is incomplete')
  }
  assertQcfV2Runs(output, expectedPages)
  return output
}

function mapVerse(
  verseKey: string,
  canonical: string[],
  qfWords: Record<string, unknown>[],
  endGlyph: string,
  translations: Map<number, string>,
  transliterations: Map<number, string[]>,
  supplements: Map<number, string>,
): RuntimeMushafWord[] {
  const rule = TOPOLOGY_RULES.get(verseKey)
  const expectedQfCount = canonical.length + (rule?.qfCount ?? 1) - (rule?.canonicalCount ?? 1)
  if (qfWords.length !== expectedQfCount) throw new Error(`QF topology changed for ${verseKey}`)
  const mapped = new Map<number, {
    glyph: string; page: number; line: number; spanEnd: number
    translation: string; transliteration: string
  }>()
  let canonicalPosition = 1
  let qfPosition = 1
  while (canonicalPosition <= canonical.length) {
    const atRule = rule?.canonicalStart === canonicalPosition && rule.qfStart === qfPosition ? rule : null
    const canonicalCount = atRule?.canonicalCount ?? 1
    const qfCount = atRule?.qfCount ?? 1
    const source = qfWords.slice(qfPosition - 1, qfPosition - 1 + qfCount)
    const page = integer(source[0]!.page_number, 'page_number')
    const line = integer(source[0]!.line_number, 'line_number')
    if (source.some((word) => Number(word.page_number) !== page || Number(word.line_number) !== line)) {
      throw new Error(`A joined QF word crosses a line at ${verseKey}`)
    }
    const ids = source.map((word) => integer(word.word_id, 'word_id'))
    mapped.set(canonicalPosition, {
      glyph: source.map((word) => string(word.text, 'text')).join(' '),
      page, line, spanEnd: canonicalPosition + canonicalCount - 1,
      translation: joinGloss(ids.map((id) => requiredMap(translations, id, 'translation'))),
      transliteration: joinGloss(ids.map((id) => supplements.get(id) ??
        single(transliterations.get(id), `QF transliteration is ambiguous or missing for word ${id}`))),
    })
    canonicalPosition += canonicalCount
    qfPosition += qfCount
  }
  if (qfPosition !== qfWords.length + 1) throw new Error(`QF topology ended early for ${verseKey}`)
  const lastOwner = Math.max(...mapped.keys())
  mapped.get(lastOwner)!.glyph += ` ${endGlyph}`
  const [surah, ayah] = verseKey.split(':').map(Number)
  const ayahPage = integer(qfWords[0]!.page_number, 'page_number')
  return canonical.map((_, index) => {
    const position = index + 1
    const value = mapped.get(position) ?? {
      glyph: '', page: 0, line: 0, spanEnd: position, translation: '', transliteration: '',
    }
    return {
      record_type: 'mushaf_word', record_key: `${verseKey}:${position}`,
      surah_id: surah!, ayah_number: ayah!, position,
      translation_en: value.translation, transliteration: value.transliteration,
      qcf_v2: value.glyph, qcf_page: value.page, qcf_line: value.line,
      qcf_span_end: value.spanEnd, ayah_page: ayahPage,
    }
  })
}

function groupedWordText(rows: Record<string, unknown>[]): Map<number, string[]> {
  const values = new Map<number, string[]>()
  for (const row of rows) {
    const id = integer(row.word_id, 'word_id')
    values.set(id, (values.get(id) ?? []).concat(string(row.text, 'text').trim()))
  }
  return values
}

function uniqueWordText(rows: Record<string, unknown>[], label: string): Map<number, string> {
  const grouped = groupedWordText(rows)
  const result = new Map<number, string>()
  for (const [id, values] of grouped) {
    if (values.length !== 1) throw new Error(`Duplicate QF ${label} word owner`)
    result.set(id, values[0]!)
  }
  return result
}

function joinGloss(parts: string[]) {
  return parts.filter(Boolean).filter((part, index, all) => index === 0 || part !== all[index - 1]).join(' ')
}

function resource(resources: StoredQfResource[], group: string, id: number) {
  const matches = resources.filter((value) => value.resourceGroup === group && value.resourceId === id)
  if (matches.length !== 1) throw new Error(`QF resource ${group}:${id} is missing`)
  return matches[0]!
}

function requiredMap(values: Map<number, string>, id: number, label: string) {
  const value = values.get(id)
  if (value == null) throw new Error(`QF ${label} is missing for word ${id}`)
  return value
}

function single(values: string[] | undefined, message: string) {
  if (values?.length !== 1) throw new Error(message)
  return values[0]!
}

function compareVerseKeys(first: string, second: string) {
  const [as, aa] = first.split(':').map(Number)
  const [bs, ba] = second.split(':').map(Number)
  return as! - bs! || aa! - ba!
}

function integer(value: unknown, name: string): number {
  if (!Number.isInteger(value)) throw new Error(`QF field ${name} is missing`)
  return Number(value)
}

function string(value: unknown, name: string): string {
  if (typeof value !== 'string') throw new Error(`QF field ${name} is missing`)
  return value
}

const QCF_V2_FIRST_CODEPOINT = 0xfc41

/** Proves that each glyph is in the contiguous run encoded by its page font. */
export function assertQcfV2Runs(records: RuntimeMushafWord[], expectedPages: readonly number[]) {
  const pages = new Map<number, RuntimeMushafWord[]>()
  for (const row of records) {
    if (!row.qcf_v2.trim()) continue
    pages.set(row.qcf_page, (pages.get(row.qcf_page) ?? []).concat(row))
  }
  for (const pageNumber of expectedPages) {
    const page = pages.get(pageNumber)
    if (!page) throw new Error(`QCF V2 page ${pageNumber} has no glyphs`)
    const codes = page
      .sort((a, b) => a.surah_id - b.surah_id || a.ayah_number - b.ayah_number || a.position - b.position)
      .flatMap((row) => [...row.qcf_v2].filter((character) => !/\s/u.test(character))
        .map((character) => character.codePointAt(0)!))
    codes.forEach((code, index) => {
      const expected = QCF_V2_FIRST_CODEPOINT + index
      if (code !== expected) throw new Error(
        `QCF V2 page ${pageNumber} glyph ${index} is U+${code.toString(16).toUpperCase()}, ` +
        `expected U+${expected.toString(16).toUpperCase()}`,
      )
    })
  }
}
